File: leetcode/medium/coin_change_2/coin_change_2.go
package main

func change(amount int, coins []int) int {
	i := 0
	for _, coin := range coins {
		if coin <= amount {
			coins[i] = coin
			i += 1
		}
	}
	coins = coins[0:i]
	tab := map[int]map[int]int{}
	return _change(amount, coins, tab)
}

func _change(amount int, coins []int, tab map[int]map[int]int) int {
	if amount == 0 {
		return 1
	}
	if amount < 0 || len(coins) == 0 {
		return 0
	}
	n := len(coins)

	m, ok := tab[n]
	if ok {
		r, ok := m[amount]
		if ok {
			return r
		}
	} else {
		m = map[int]int{}
		tab[n] = m
	}
	r := 0
	for i, coin := range coins {
		r += _change(amount-coin, coins[0:i+1], tab)
	}
	m[amount] = r
	return r
}

File: leetcode/medium/coin_change_2/coin_change_2.py

class Solution:
    def change(self, amount, coins):
        coins = [coin for coin in coins if coin <= amount]
        tab = {}
        def _change(amount, n):
            if amount == 0:
                return 1
    
            if amount < 0 or n == 0:
                return 0
        
            # amount > 0 and n > 0
            key = (amount, n)
            r = tab.get(key)
            if r is not None:
                return r
    
            r = 0
            for i in range(n):
                r += _change(amount - coins[i], i+1)
    
            tab[key] = r    
            return r

        return _change(amount, len(coins))

